### PR TITLE
Adds full configuration match possibility for option availability check

### LINKED ISSUE #4831: Function "isOptionAvailable()" should have option "fullMatch"
## Current behavior

the function core\modules\catalog\helpers\variant\isOptionAvailable.ts uses the function findConfigurableVariant()

This function will ALWAYS return a variant... not really the expected behavior

## Expected behavior

The function findConfigurableVariant() should also have an option "fullMatch" true/false... with default "false" to not have a breaking change...

## Version of Vue Storefront

* [x] Vue Storefront
* [ ] Vue Storefront Next

## Can you handle fixing this bug by yourself?
* [x] YES
* [ ] NO

## Which [Release Cycle](https://docs.vuestorefront.io/guide/basics/release-cycle.html) state this refers to? Info for developer. (doesn't apply to Next)
Pick one option.

* [x] This is a bug report for test version on https://test.storefrontcloud.io - In this case Developer should create branch from `develop` branch and create Pull Request `2. Feature / Improvement` back to `develop`.
* [x] This is a bug report for current Release Candidate version on https://next.storefrontcloud.io - In this case Developer should create branch from `release` branch and create Pull Request `3. Stabilisation fix` back to `release`.
* [x] This is a bug report for current Stable version on https://demo.storefrontcloud.io and should be placed in next stable version hotfix - In this case Developer should create branch from `hotfix` or `master` branch and create Pull Request `4. Hotfix` back to `hotfix`.

## Environment details

* Browser: 
* OS: 
* Node: 
* Code Version:  1.12.3

### PR DESCRIPTION
**Pull request by simonmaass**
### Related Issues

closes #4831

### Short Description and Why It's Useful

the expected behaviour of isOptionAvailable() suggests that if you give it a configuration that does not exists it will return 'null' and not ALWAYS return a variation... therefore added 'fullMatch' parameter

### Which Environment This Relates To
Check your case. In case of any doubts please read about [Release Cycle](https://docs.vuestorefront.io/guide/basics/release-cycle.html)

- [ ] Test version (https://test.storefrontcloud.io) - this is a new feature or improvement for Vue Storefront. I've created branch from `develop` branch and want to merge it back to `develop`
- [ ] RC version (https://next.storefrontcloud.io) - this is a stabilisation fix for Release Candidate of Vue Storefront. I've created branch from `release` branch and want to merge it back to `release`
- [ ] Stable version (https://demo.storefrontcloud.io) - this is an important fix for current stable version. I've created branch from `hotfix` or `master` branch and want to merge it back to `hotfix`

### Upgrade Notes and Changelog
- [x] No upgrade steps required (100% backward compatibility and no breaking changes)
- [ ] I've updated the [Upgrade notes](https://github.com/DivanteLtd/vue-storefront/blob/develop/docs/guide/upgrade-notes/README.md) and [Changelog](https://github.com/DivanteLtd/vue-storefront/blob/develop/CHANGELOG.md) on how to port existing Vue Storefront sites with this new feature

**IMPORTANT NOTICE** - Remember to update `CHANGELOG.md` with description of your change

### Contribution and Currently Important Rules Acceptance

- [ ] I read and followed [contribution rules](https://github.com/DivanteLtd/vue-storefront/blob/master/CONTRIBUTING.md)

### CODE REVIEW

File: core/modules/catalog/helpers/variant/checkFullConfigurationMatch.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+  const configProperties = Object.keys(omit(configuration, ['price']))
+  const countMatchedConfigProperties = configProperties
+    .map(configProperty => {
+      const variantPropertyId = variant[configProperty]
+      if (configuration[configProperty] === null) {
+        return false
+      }
+
+      return [].concat(configuration[configProperty])
+        .map(f => typeof f === 'object' ? toString(f.id) : f)
```
Comment (gibkigonzo) [inline core/modules/catalog/helpers/variant/checkFullConfigurationMatch.ts:18]: if `f` is **not** an object then you don't need to change it to string?

File: core/modules/catalog/helpers/variant/checkFullConfigurationMatch.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+/**
+ * Checks if full coniguration matches for specific variant
+ */
+export default function checkFullConfigurationMatch (configuration, variant): boolean {
+  if (!variant || !configuration) return false
+  const configProperties = Object.keys(omit(configuration, ['price']))
+  const countMatchedConfigProperties = configProperties
+    .map(configProperty => {
+      const variantPropertyId = variant[configProperty]
+      if (configuration[configProperty] === null) {
```
Comment (gibkigonzo) [inline core/modules/catalog/helpers/variant/checkFullConfigurationMatch.ts:13]: wdyt, about checking if it not exist? `if (!configuration[configProperty]) return`

File: core/modules/catalog/helpers/variant/checkFullConfigurationMatch.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
+import toString from 'lodash/toString'
+import omit from 'lodash/omit'
+
+/**
+ * Checks if full coniguration matches for specific variant
+ */
+export default function checkFullConfigurationMatch (configuration, variant): boolean {
+  if (!variant || !configuration) return false
+  const configProperties = Object.keys(omit(configuration, ['price']))
```
Comment (gibkigonzo) [inline core/modules/catalog/helpers/variant/checkFullConfigurationMatch.ts:9]: why you make omit for price? 🤔 just curious, can price be set as configuration property?

File: core/modules/catalog/helpers/variant/isOptionAvailable.ts [hunk 4]
```diff
@@ -3,7 +3,7 @@ import findConfigurableVariant from './findConfigurableVariant'
 /**
  * Checks if variant with specific configuration exist
  */
-export default function isOptionAvailable (context, { product, configuration }): boolean {
-  const variant = findConfigurableVariant({ product: product, configuration: configuration, availabilityCheck: true })
+export default function isOptionAvailable (context, { product, configuration, fullMatch = false }): boolean {
+  const variant = findConfigurableVariant({ product: product, configuration: configuration, availabilityCheck: true, fullMatch: fullMatch })
```
Comment (gibkigonzo) [inline core/modules/catalog/helpers/variant/isOptionAvailable.ts:7]: can you change it?
```
const variant = findConfigurableVariant({
  product,
  configuration,
  availabilityCheck: true,
  fullMatch
})
```

**Review by gibkigonzo — COMMENTED** [review]
looks good, I've added some comments 👍

File: core/modules/catalog/helpers/variant/isOptionAvailable.ts (hunk 4 above)
Comment (Fifciu) [reply core/modules/catalog/helpers/variant/isOptionAvailable.ts:7]: Or maybe even?:
```js
const variant = findConfigurableVariant({
  product,
  configuration,
  fullMatch,
  availabilityCheck: true
})
```

To elegantly separate properties with the provided value from shortened ones

**Review by Fifciu — CHANGES_REQUESTED** [review]

**Comment by simonmaass** [comment]
about all the comments... totally agree... not really my code thou...

i just copied the code from core\modules\catalog\helpers\variant\getConfigurationMatchLevel.ts....

Should probably be smart to also change this method then... wdyt?

**Comment by gibkigonzo** [comment]
@simonmaass haha you are right :D I just looked on `core\modules\catalog\helpers\variant\getConfigurationMatchLevel.ts`
If you can, please apply suggestions also to `getConfigurationMatchLevel`

**Comment by bloodf** [comment]
This is closed due to inactivity.

If you want your PR to be re-opened, please update your files, check for any conflicts.

If your PR is for Vue Storefront 1 target the new repository at https://github.com/vuestorefront/vue-storefront-1

**Pull request closed without merging**